Fixed Tests for Turtle Singer and Synth Utils

ac-mmi opened the pull request:
**This PR fixes issues with the tests for TurtleSinger, SynthUtils, and VolumeActions which i saw in my previous PR. There were some errors popping up due to undefined properties and some inconsistent mocks, so I cleaned that up.**

In turtle-singer.test.js, I fixed TypeErrors by properly setting up blockList and connections. I also improved the mocking for Tone.Panner and updated test cases to cover more edge cases. In synthutils.test.js, I resolved issues with mocks causing errors and initialized activity.logo.synth correctly to avoid undefined errors. 

In volumeactions.test.js, I fixed undefined connections by setting up blockList as follows:
```
blockList: {
       mockBlk: {
         connections: ['mockConnection1', 'mockConnection2']
         }
}
```
I ran the `npm test` command, and all the tests passed.

![Screenshot (2585)](https://github.com/user-attachments/assets/5b56da4a-197a-48ad-b2b6-3e6fef2c4e99)

omsuneri commented:
@ac-mmi try to look at the #4422 as this PR resolves all the errors of volumeaction also you are not supposed to make changes in the root file also try to folow our guide for implementing tests also i observed that you had changed multiple po files which need to be done in another PR

ac-mmi commented:
@omsuneri i have removed those po files. The changes that i have done in the root file in `turtle-singer.js` was this line
`const lastConnection = activity.logo.blockList[blk].connections.slice(-1)[0] || null;`
where earlier it used last() function but when i ran test case i got last() as undefined function. 

![Screenshot (2586)](https://github.com/user-attachments/assets/81e13dfb-9dbf-487f-87cb-dd13ce1847c2)

ac-mmi commented:
https://github.com/sugarlabs/musicblocks/pull/4422#issuecomment-2671840720
@walterbender I added the following to activityMock to ensure that setMasterVolume receives the first and last connection arguments correctly:

```
blockList: {
    mockBlk: {
        connections: ['mockConnection1', 'mockConnection2']
    }
}
```
I also updated the tests to ensure setMasterVolumeand setSynthVolume are called with the correct arguments.

```
test('should set master volume correctly', () => {
    Singer.setMasterVolume(logoMock, 50, 'mockBlk');
    expect(logoMock.synth.setMasterVolume).toHaveBeenCalledWith(50, 'mockConnection1', 'mockConnection2');
});

test('should set synth volume correctly', () => {
    Singer.setSynthVolume(logoMock, turtleMock, 'noise1', 80, 'mockBlk');
    expect(logoMock.synth.setVolume).toHaveBeenCalledWith(turtleMock, 'noise1', 80 / 25, 'mockBlk');
});
```

omsuneri commented:
@ac-mmi try to mock the last as global.last = jest.fn((array) => array[array.length - 1]);

omsuneri commented:
@ac-mmi also try to remove the volumaction part as well as those errors are removed in the #4422 and also there i had added many new tests so kindly please do that tooo

ac-mmi commented:
@omsuneri Done

omsuneri commented:
@ac-mmi no issues now this looks good !!

walterbender commented on js/utils/synthutils.js:
```diff
@@ -1349,6 +1349,7 @@ function Synth() {
         synthOptions = validateAndSetParams(synthOptions, params);
 
         let tempSynth;
+        console.log(sourceName,'AMANANANAN');
```
please remove this and other debugging console messages.

walterbender commented:
please remove the unnecessary comments, such as // New additon

walterbender commented on js/__tests__/turtle-singer.test.js:
```diff
@@ -54,6 +56,12 @@ describe('Singer Class', () => {
+                // Add blockList here
+                blockList: {
+                    mockBlk: {
+                        connections: ['mockConnection1', 'mockConnection2']
```
these should really be ints, but it doesn't really matter as long as they are not null.

ac-mmi replied:
so should i change them or not

walterbender replied:
Maybe changes them to [0, 0] and then change it in the test too.

ac-mmi replied:
Done

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
         });
         it("it should creates a amsynth based on the specified parameters, either using samples, built-in synths, or custom synths", () => {
-            const instrumentName = "guitar"
+            let instrumentName = "guitar"
```
This will work, but it would make the test less confusing if the instrumentName was an example of an AM Synth. "poly" for example.

ac-mmi replied:
Done

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
             CUSTOMSAMPLES['pianoC4'] = "pianoC4";
             CUSTOMSAMPLES['drumKick'] = "drumKick";
-            const instrumentName = "guitar"
+            let instrumentName = "guitar"
```
maybe use "piano"?

walterbender replied:
But isn't that also true for guitar?

ac-mmi replied:
nope

walterbender replied:
Why doesn't guitar, which is also a sample get initialized as a Sampler?

ac-mmi replied:
i added some mock functions in synth which weren't present like start,stop,triggerAttack and triggerRelease which were earlier posing problem in the test cases

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
-            const instrumentName = "guitar"
-            const sourceName = "http://testing.com"
+            let instrumentName = "guitar"
+            let sourceName = "http://testing.com"
```
Maybe a comment to explain why this URL?

ac-mmi replied:
done

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
         it("it should creates a CUSTOMSAMPLES based on the specified parameters, either using samples, built-in synths, or custom synths", () => {
-            const instrumentName = "guitar"
-            const sourceName = "http://testing.com"
+            let instrumentName = "guitar"
```
Perhaps a drum name? or something suggesting a relationship to the faux URL?

ac-mmi replied:
done

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
-            const instrumentName = "guitar"
-            const sourceName = "file://testing.jpg"
+            let instrumentName = "guitar"
+            let sourceName = "file://testing.jpg"
```
why jpg? Maybe the faux name should be .wav?

ac-mmi replied:
@walterbender yes you are right

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -209,36 +209,36 @@ describe("Utility Functions (logic-only)", () => {
         it("it should creates a CUSTOMSAMPLES based on the specified parameters, either using samples, built-in synths, or custom synths", () => {
-            const instrumentName = "guitar"
-            const sourceName = "drum"
+            let instrumentName = "guitar"
```
a drum name? "snare"?

ac-mmi replied:
done

walterbender commented on js/utils/__tests__/synthutils.test.js:
```diff
@@ -1024,9 +1024,14 @@ describe("Utility Functions (logic-only)", () => {
         it("it should creates a synth based on the user's input in the 'Timbre' clamp, handling race conditions with the samples loader.", () => {
-            expect(createSynth("turtle1", "piano", "voiceSample1", {})).toBe(undefined);
+            const turtle = "turtle1";  // Use const or let
+            const instrumentName = "piano";  // Localize declaration
```
maybe "voice"? "voice recording"?

ac-mmi replied:
Done

walterbender commented:
What are the .po.tmp files for?

ac-mmi commented:
@walterbender The `.po.tmp` files are simplified versions of the `.po` files with just the **msgid** and **msgstr** values, without any comments or extra details. I noticed this when working on PR #4433 , where I fixed an issue with **"gong"** word in `quz.po`. It had an empty `msgstr`, and the .tmp file stopped processing at that point. I think it’s meant to make processing easier but might get stuck on problematic entries.

walterbender commented:
Please make any changes to PO files in a separate PR. Also, the comments in the PO files are there to help the translators.

ac-mmi commented:
@walterbender, I noticed the PR got closed, and I’m not sure why. Could you let me know the reason? I’d like to understand so I can learn from it and improve. If possible, I’d like to continue working on this. Thanks!

walterbender commented:
It was closed by accident. Sorry about that. Fat fingers today.

ac-mmi commented:
@walterbender no problem
